sound/soc/sof/topology.c

```diff
@@ -1996,12 +2000,28 @@ static int sof_manifest(struct snd_soc_component *scomp, int index,
 		}
 	}
 
-	/* copy nhlt blob after that should be after ABI data */
+	/* we have more data in manifest than just abi number, so parse it */
 	if (size > SOF_TPLG_ABI_SIZE) {
-		sdev->nhlt_blob = devm_kmemdup(sdev->dev, &man->priv.data[SOF_TPLG_ABI_SIZE],
-					       size - SOF_TPLG_ABI_SIZE, GFP_KERNEL);
-		if (!sdev->nhlt_blob)
-			return -ENOMEM;
+		manifest = (struct sof_manifest *)man->priv.data;
+		manifest_data = manifest->items;
+		len_check = SOF_TPLG_ABI_SIZE + sizeof(manifest->count);
+		for (i = 0; i < manifest->count; i++) {
+			len_check += sizeof(struct sof_manifest_data) + manifest_data->size;
+			if (len_check > size)
+				return -EINVAL;
+			switch (manifest_data->type) {
+			case SOF_MANIFEST_DATA_TYPE_NHLT:
+				sdev->nhlt_blob = devm_kmemdup(sdev->dev, manifest_data->data,
```

Comment:
can there be more than 1 NHLT blobs? what happens if there are?

Reply:
I think it's the same blob/table confusion we've discussed before. The blob is only valid as a 'Specific_Config' within the scope of a format, and there can be multiple formats for multiple endoints. there is ONE table however.

Reply:
right I think in this case we should check is sdev->nhlt_blob is not NULL and return error

Reply:
and we should rename nhlt_blob to nhlt_table right?

Reply:
I think nhlt_blob (or just nhlt) would be best. So with nhlt I understand superstructure that contains all endpoint descriptions, so possibly many vendor specific blobs. But again I'm not sure about acpi table terminology, I'm fine with all names if we just agree.

Reply:
'nhlt' is enough, no need for a blob qualifier that's misleading, and no need to add table when the 't' stands for table already.

Reply:
changed to just nhlt